### PR TITLE
Heap based router queue

### PR DESCRIPTION
**Pull request by mrs303**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Review by szutenberg — APPROVED** [review]

File: router/src/queue.rs [hunk 1]
```diff
@@ -132,11 +135,102 @@ async fn queue_task(
...
+            self.overdue_entries.pop()
+        } else {
+            self.regular_entries.pop()
+        }
+    }
+
+    fn is_empty(&self) -> bool {
+        self.regular_entries.is_empty() && self.overdue_entries.is_empty()
+    }
+    fn len(&self) -> usize {
```
Comment (kdamaszk) [inline router/src/queue.rs:226]: new line

File: router/src/queue.rs [hunk 2]
```diff
@@ -132,11 +135,102 @@ async fn queue_task(
...
+        self.regular_entries.len() + self.overdue_entries.len()
+    }
+}
 /// Queue State
```
Comment (kdamaszk) [inline router/src/queue.rs:231]: new line

File: router/src/queue.rs [hunk 3]
```diff
@@ -221,7 +323,10 @@ impl State {
         let mut decode_tokens: u32 = 0;
 
         // Pop entries starting from the front of the queue
-        while let Some((id, mut entry)) = self.entries.pop_front() {
+        while let Some(id_entry) = self.entries.pop() {
```
Comment (kdamaszk) [inline router/src/queue.rs]: Why default implementation was wrong here? Less changes mean easier merge with upstream later.

File: router/src/queue.rs [hunk 4]
```diff
@@ -166,10 +260,16 @@ impl State {
         max_input_length: u32,
         max_total_tokens: u32,
         block_size: u32,
-        window_size: Option<u32>
+        window_size: Option<u32>,
     ) -> Self {
+        let default_threshold: u64 = 120;
+        let threshold: u64 = match env::var("QUEUE_THREASHOLD_MS") {
```
Comment (kdamaszk) [inline router/src/queue.rs]: typo

Comment (kdamaszk) [reply router/src/queue.rs]: add flag to README

**Review by kdamaszk — CHANGES_REQUESTED** [review]

**Review by kdamaszk — APPROVED** [review]

**Pull request merged**